Add test and fix signature off-by-one

mhoogendoorn opened the pull request:
The `fsharp/signature` extension also seems to be a victim of the 'Great Offset Shift of 2021' (see #767, #774, #771).

The tests are basically the first tooltip tests from CoreTests.fs.  Not sure what is preferred, to have them there because they share logic (e.g. `server`), or if they should be in ExtensionTests because they test a custom endpoint.  Currently in the latter, should it have a separate TestCases file instead of reusing the TestCases/Tooltips.fs file?

(a quick search shows a few more places where `Lexer.findLongIdents(pos.Column - 1, lineStr)` is used, likely they are affected as well).

mhoogendoorn committed:
Add test and fix signature off-by-one

baronfel commented:
This is a wonderful surprise :) Thank you for taking the time to investigate and write some tests!  Yes, the Great Offset Shift is having reprecussions to this day, not the least of which was moving all text reading over to a less-error-prone model! (see ISourceText, etc).

You're likely correct about these follow-on places that use `pos.Column - 1`, I've had them as kind of a mental bucket list to add tests to and pin behavior down, but hadn't gotten around to them yet!

Regarding placement/naming of the tests, the name `ExtensionsTests` is meant to evoke the fact that Fake/FSharpLint/Fantomas are all external integrations, etc.  I think the ideal case would be a new file for this endpoint specifically, we try to limit the blast radius by having 'isolated' test files as much as possible.

baronfel commented on test/FsAutoComplete.Tests.Lsp/ExtensionsTests.fs:
```diff
@@ -375,3 +375,50 @@ let analyzerTests state =
+  testSequenced <|
+    testList "signature evaluation" [
+      testList "tests" [
+        verifySignature 0 4 "val arrayOfTuples : (int * int) []"  // verify that even first letter of the signature triggers correctly
```
what I'd really love to be able to get to at some point with many of these tests is testing a range of similar positions start to finish, eg the entire length of an identifier. That would be the most durable test, but in lieu of that can you also add one that uses the end of the identifier as well? just to cover the extremes.

mhoogendoorn replied:
Changed it so it tests all positions between a given start and end. I've kept it in a single testCaseAsync so a 12 character identiefer does not produce 12 separate tests.  If it fails it will still say at which character it failed first.

baronfel commented on test/FsAutoComplete.Tests.Lsp/ExtensionsTests.fs:
```diff
@@ -375,3 +375,50 @@ let analyzerTests state =
+    testCaseAsync (sprintf "fsharp/signature for line %d character %d should be '%s'" line character expectedSignature) (async {
+      let! server, scriptPath = server
+      let pos: TextDocumentPositionParams = {
+        TextDocument =  { Uri = sprintf "file://%s" scriptPath }
```
There's a utility we use, `Path.FilePathToUri` that we use for these cases to fully-encode the path, because there's some cross-platform nastiness around converting file paths to file URIs. Please use that here as well.

baronfel reviewed (changes requested):
Great work overall! Just a couple requests here.

baronfel commented:
Regarding the test results here, the only one you really need to worry about is the macos one. There are some core issues we have to address on the Windows tests (a combination of case-sensitive path comparisons as well as some _serious_ flakiness on the windows runners in GitHub Actions), as well as the ubuntu tests (unknown/undiagnosed hangs).

mhoogendoorn committed:
Adjustments to fsharp/signature tests

Use a separate testcase file for this endpoint.  Use `FilePathToUri`
instead of manually creating the uri.  Test all positions of the
identifier in a single `testCaseAsync`.

baronfel commented:
Thanks, this is a great change 👍. There are a couple other fixes I want to try and land this weekend, then I'll do a release including this change.
